### PR TITLE
Issue#1545 deal list

### PR DESCRIPTION
**Pull request by ioaNikas**
- Create deal list with custom table.
- Add new rules for distributionDeals subcollections.
- Add new index for distributionDeals collectionGroup query.

### CODE REVIEW

File: apps/catalog/catalog/src/app/dashboard/deal/list/list.component.html [hunk 1]
```diff
@@ -1,4 +1,17 @@
...
-  <a [routerLink]="[contract.id]">{{ contract.id }}</a>
-</article>
+<ng-container *ngIf="contracts$ | async as contracts">
+  <h1>Offers & Deals</h1>
+
+  <mat-tab-group>
+    <mat-tab
+      *ngFor="let tab of contractTabs"
+      [label]="getLabel(tab.name, filterByStatus(contracts, tab.statuses))">
+      <contract-list
```
Comment (GrandSchtroumpf) [inline apps/catalog/catalog/src/app/dashboard/deal/list/list.component.html]: Consider using a lazy-loaded solution for your tabs to only load content when you click on it : 
https://material.angular.io/components/tabs/overview#lazy-loading

File: apps/catalog/catalog/src/app/dashboard/deal/list/list.component.ts [hunk 2]
```diff
@@ -8,8 +42,24 @@ import { ContractQuery } from '@blockframes/contract/contract/+state';
...
 
   constructor(private contractQuery: ContractQuery) {}
+
+  public getLabel(label: string, contracts: Contract[]) {
+    return `${label} (${contracts.length.toString()})`;
+  }
+
+  public filterByStatus(contracts: Contract[], statuses?: string[]) {
+    return statuses
+      ? contracts.filter(contract => statuses.includes(contract.versions[getLastVersionIndex(contract)].status))
```
Comment (GrandSchtroumpf) [inline apps/catalog/catalog/src/app/dashboard/deal/list/list.component.ts]: Do you really think we can read that easily ?
```typescript
if (statuses) {
  const lastVersionStatus = contract => contract.versions[getLastVersionIndex(contract)].status;
  return contracts.filter(contract => statuses.includes(lastVersionStatus))
}
return contracts;
```

File: libs/contract/src/lib/contract/list/contract-list.component.html [hunk 3]
```diff
@@ -0,0 +1,99 @@
+<section *ngIf="displayedColumns$ | async as displayedColumns">
```
Comment (GrandSchtroumpf) [inline libs/contract/src/lib/contract/list/contract-list.component.html]: use ng-container

File: apps/catalog/catalog/src/app/dashboard/deal/list/list.component.ts (hunk 2 above)
Comment (GrandSchtroumpf) [reply apps/catalog/catalog/src/app/dashboard/deal/list/list.component.ts]: And I'm supprise we don't already have a `getLastVersion(contract: Contract)` pure function somewhere that deal with this.

File: libs/movie/src/lib/distribution-deals/+state/distribution-deal.service.ts [hunk 4]
```diff
@@ -112,13 +132,16 @@ export class DistributionDealService extends CollectionService<DistributionDealS
...
+   * @param contractVersion
    */
-  public getDistributionDealsTerritories(deals: DistributionDeal[]): string[] {
-    const territories = deals.map(deal => deal.territory.map(territory => territory));
-    const flattenedTerritories = territories.reduce((acc, nestedArray) => acc.concat(nestedArray));
-    return flattenedTerritories;
+  public getTerritoriesFromContract(contractVersion: ContractVersion) {
+    const dealIds: string[] = [];
+    for (const title of Object.values(contractVersion.titles)) {
+      dealIds.push(...title.distributionDealIds);
```
Comment (GrandSchtroumpf) [inline libs/movie/src/lib/distribution-deals/+state/distribution-deal.service.ts]: use concat for merging two array : `dealIds.concat(title.distributionDealIds)`

File: libs/contract/src/lib/contract/+state/contract.model.ts [hunk 5]
```diff
@@ -198,7 +197,7 @@ export function createLegalDocument(
 export function createContractFromFirestore(contract: ContractWithTimeStamp): Contract {
   return {
     ...contract,
-    signDate: (contract.signDate instanceof Date) ? contract.signDate : contract.signDate.toDate(),
+    signDate: isTimestamp(contract.signDate) ? contract.signDate.toDate() : contract.signDate,
```
Comment (GrandSchtroumpf) [inline libs/contract/src/lib/contract/+state/contract.model.ts]: You could have a function `toDate(date: Timestamp | Date)` if you end up doing this several time

File: apps/catalog/catalog/src/app/dashboard/title/list/list.component.ts [hunk 6]
```diff
@@ -15,12 +16,12 @@ interface TitleView {
...
 
 function createTitleView(movie: Movie, contracts: Contract[]): TitleView {
-  const ownContracts = contracts.filter(c => c.versions[getLastVersionIndex(c)].titles[movie.id]);
+  const ownContracts = contracts.filter(c => getContractLastVersion(c).titles[movie.id]);
   return {
     title: movie.main.title.international,
     view: 'View',
     sales: ownContracts.length,
-    receipt: ownContracts.reduce((sum, contract) => sum + contract.versions[getLastVersionIndex(contract)].titles[movie.id].price.amount, 0),
+    receipt: ownContracts.reduce((sum, contract) => sum + getContractLastVersion(contract).titles[movie.id].price.amount, 0),
```
[the remark this answers is not in the document]
Comment (ioaNikas) [reply apps/catalog/catalog/src/app/dashboard/title/list/list.component.ts]: Indeed, my next PR will include a function called `getMovieReceipt` to do this stuff only on validated contracts (status === paid, waiting for signature or accepted).

**Pull request merged**